### PR TITLE
testing on transforms

### PR DESCRIPTION
**Pull request by JiazhengZhu**

### CODE REVIEW

File: test/transforms.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+def img():
+    return Image.open('test/4.2.03.tiff').convert('L')
+
+# flip is concordant with PIL flip
+
+
+def test_flip(img):
+    output = _flip()(img)
+    ground = ImageOps.flip(ImageOps.mirror(img))
+    assert np.allclose(np.array(output), asarray(
```
Comment (tdfirth) [inline test/transforms.py:23]: This is really cool, quite a unique solution. My understanding is that you have found an equivalent operation in Pillow and you testing that they work the same as each other?

If you're interested, check out how I went about it here https://github.com/tdfirth/ser/blob/4ce627fc76bfef8ce5b85b05e53a11350f9cda94/tests/test_transforms.py#L27

Rather than loading a real image I went down the path of creating a small tensor manually. There are pros and cons to both approaches but generally if you can make a little bit of fake data it's easier to orchestrate and add coverage for new edge cases.

File: test/transforms.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+# normalising gives correct mean and std
+
+
+def test_norm(img):
+    n = _normalize()
+    output = n(ToTensor()(img))
+    pixels = asarray(output)
+    ground = (np.array(img)/255 - 0.5)/0.5
+    assert np.allclose(pixels, ground, atol=1e-4)
```
Comment (tdfirth) [inline test/transforms.py:34]: Also very clever, quite neat this :). Again see mine for comparison if you're interested.

File: ser/transforms.py [hunk 3]
```diff
@@ -1,29 +1,42 @@
...
     return torch_transforms.Compose(
         [
             torch_transforms.RandomHorizontalFlip(p=1.0),
             torch_transforms.RandomVerticalFlip(p=1.0),
+            torch_transforms.ToTensor(),
         ]
     )
+
+
+def transforms_by_option(lst):
```
Comment (tdfirth) [inline ser/transforms.py:32]: This is great, love being able to specify the name of the transform, but I think you could have simplified it a little. Maybe something like this is easier to follow?

```
TRANSFORMS = {
    'flip': _flip,
    'normalise': _normalize,
}

def transforms_by_option(transforms):
    result = []
    for t in transforms:
        if t not in TRANSFORMS:
            raise NameError(f'Unsupported transformation: {t}')
        result.append(TRANSFORMS[t])
    return _transforms(*result)
```

You unit tested this which was cool. Once the unit test is there I find it provides quite a good opportunity to revisit the implementation and iterate on it a couple of times before moving on.

**Review by tdfirth — COMMENTED** [review]
I really enjoyed reading your PR, I love how you've approached things independently and come up with unique solutions. Some bits in there are clever I think.

My only 'review comment' really was on that function for looking up the transformation function from user input. It's a good idea but the implementation was just a bit more complex than it needed to be.

In general, I think you were great this week, you're clearly a smart guy and I think you have a healthy dose of scepticism for some of the stuff we covered. This could just be from past experience where things didn't work for you but either way, don't change that. I hope I got the point across that there are no magic bullets and that engineering is mostly just about discipline, trade-offs, pragmatism, etc. Techniques and tools play a role but are useless without that context.

That said, I hope you found the week useful and that you apply some of the ideas in the future! Do fill out the feedback form if you get a chance, especially if you think some things could be improved. 

Good luck with the DPhil :).
